sso-url-helper/src/main/java/it/niedermann/nextcloud/sso/urlhelper/UrlHelper.kt

```diff
@@ -0,0 +1,132 @@
+val Context.fallbackPreviewSize get() = FallbackPreviewSize(
+    width = resources.displayMetrics.widthPixels,
+    height = resources.displayMetrics.heightPixels,
+)
+
+/**
+ * Generates an [URL] object from the given [model]
+ * @throws [IllegalArgumentException] in case the [model] can not be processed or the [ssoAccount] is not valid for the [model]
+ */
+fun getAbsoluteUrl(ssoAccount: SingleSignOnAccount, model: String, fallbackPreviewSize: FallbackPreviewSize): URL {
+    return try {
+        // Check whether this is a complete URL, will throw an MalformedURLException if not
+        val url = URL(model)
+        // Verify it starts with the given ssoAccount
+        if (model.startsWith(ssoAccount.url)) {
+            rewriteSpecialURLs(ssoAccount, url, fallbackPreviewSize).orElse(url)
+        } else {
+            throw IllegalArgumentException("Given ${SingleSignOnAccount::class.java.simpleName} does not match the URL (${ssoAccount.url} vs. ${model}). Pass correct ${SingleSignOnAccount::class.java.simpleName} or use correct URL (or a plain ${String::class.java.simpleName}) to try fetching with the current ${SingleSignOnAccount::class.java.simpleName} stored in ${SingleAccountHelper::class.java.simpleName}.")
+        }
+    } catch (e: MalformedURLException) {
+        if (model.startsWith("/")) {
+            // This might be an absolute path instead of an URL, so prepend the URL of the ssoAccount, but be aware of accounts which are located in a sub directory!
+            val url = URL(ssoAccount.url.substring(0, ssoAccount.url.length - URL(ssoAccount.url).path.length) + model)
+            rewriteSpecialURLs(ssoAccount, url, fallbackPreviewSize).orElse(url)
+        } else {
+            throw IllegalArgumentException("URL must be absolute (starting with protocol and host or with a slash character).")
+        }
+    }
+}
+
+/**
+ * @return a rewritten [URL] if something special has been detected, like Share IDs, File IDs or avatars. [Optional.empty] otherwise.
+ */
+private fun rewriteSpecialURLs(ssoAccount: SingleSignOnAccount, url: URL, fallbackPreviewSize: FallbackPreviewSize): Optional<URL> {
+    // Exclude potential sub directory from url path (if Nextcloud instance is hosted at https://example.com/nextcloud)
+    val pathStartingFromNextcloudRoot = if (url.query == null) {
+        url.toString().substring(ssoAccount.url.length)
+    } else {
+        val urlStringWithoutLeadingAccount = url.toString().substring(ssoAccount.url.length)
+        urlStringWithoutLeadingAccount.substring(0, urlStringWithoutLeadingAccount.length - url.query.length - 1)
+    }
+
+    val fileId = REGEX_FILE_ID.find(pathStartingFromNextcloudRoot)?.groupValues?.get(2)
+    if (fileId != null) {
+        return if (url.query == null) {
+            Optional.of(URL("${ssoAccount.url}/index.php/core/preview?fileId=${fileId}&x=${fallbackPreviewSize.width}&y=${fallbackPreviewSize.height}&a=true"))
```

Comment:
Let me know if you would prefer passing the whole context. I just thought it would be more flexible this way.

Reply:
That's fine :+1: